gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -404,10 +414,59 @@ public void delete(URI path, boolean recursive) throws IOException {
               : listFileInfoForPrefixPage(
                       fileInfo.getPath(), DELETE_RENAME_LIST_OPTIONS, /* pageToken= */ null)
                   .getItems();
-      if (!itemsToDelete.isEmpty() && !recursive) {
+
+      if (isHnBucket) {
+        /**
+         * Get list of folders in path if the bucket is HN enabled bucket For recursive delete, get
+         * all folder resources. For non-recursive delete, only get maximum of 1 folder resources
+         * other than the path.
+         */
+        listOfFolders =
```

Comment:
It is possible to do the listing asynchronously, I guess.

Reply:
Added a todo